lib/Semantics/check-declarations.cpp

```diff
@@ -707,6 +707,24 @@ void CheckHelper::CheckDerivedType(
         }
       }
     }
+    DerivedTypeSpec derived{symbol.name(), symbol};
+    derived.set_scope(*symbol.scope());
+    if (FindCoarrayUltimateComponent(derived) && // C736
+        !(parentDerived && FindCoarrayUltimateComponent(*parentDerived))) {
+      messages_.Say(
+          "Type '%s' has a coarray ultimate component so the type it extends "
+          "('%s') must have a coarray ultimate component"_err_en_US,
+          symbol.name(), parentDerived->typeSymbol().name());
+    }
+    if (FindEventOrLockPotentialComponent(derived) && // C737
+        !(FindEventOrLockPotentialComponent(*parentDerived) ||
+            IsEventTypeOrLockType(parentDerived))) {
+      messages_.Say(
+          "Since type '%s' has an EVENT_TYPE or LOCK_TYPE component, the type"
```

Comment:
The wording is strictly correct here but might be misleading; the parent type does not have to have an event or lock component itself, but can inherit one from its parent.  What's required by Fortran is that the ultimate parent type (the "oldest" ancestor, the one without an `EXTENDS`) must have (or be) an event or lock.

Suppose we have some non-extended derived type X, and that Y extends X, and that Z extends Y.  And suppose that Z contains an event or lock.  C737 requires that Y must have an event or lock, directly or via inheritence.  But if Y has an event or lock itself, then C737 applies to Y as well, and Z must have (or be) an event or lock.  If Y does not have an event or lock, then C737 on Z requires that X must have (or be) an event or lock.  So it always works out that the deepest parent derived type ends up having to have (or be) an event or lock, and we might as well make that clear in our implementation and our error messages, even if the Standard chooses to be gratuitously confusing.

This applies to coarray component checks too, I think.

(Please check my reasoning, maybe I've missed something.)

Reply:
Thanks, Peter.  I'll enhance the test to add a more deeply nested test case and check the behavior of other compilers.

Reply:
I think that you want to discover the *most* deeply nested ancestor type and test that, rather than the immediate parent, and adjust the message to complain about the constraint in terms of the most deeply nested ancestor type.

Reply:
Also, I'll reword the message to be more descriptive of the actual requirement.

Reply:
Correct me if I'm wrong, but constraint C736, which applies to coarrays, requires that the parent type have a `coarray ultimate component`, whereas constraint C737, which applies to `LOCK_TYPE` and `EVENT_TYPE` states that the parent type must have a `potential subobject component`.  If I understand things correctly, this means that the parent must directly contain a coarray component.  This is the interpretation of GNU Fortran, which is the only compiler I could find that both supports coarrays and enforces C736 at all.

Reply:
Direct components and ultimate components and potential subobject components are three distinct concepts.  See 7.5.1.  It is true that C736 is defined in terms of ultimate components and that C737 is defined in terms of potential subobject components.

In short: Direct components are those that are always *present* in every instance.  Ultimate components are the direct components, plus pointers and allocatables,, but not the contents of those pointers and allocatables.  Potential subobject components are the direct components, plus the potential subobject components of allocatables -- they are all of the storage *owned* by the instance.

Reply:
I take it all back.  It looks like your interpretation of coarrays is correct.  I'll update the error message and add a test.